### ISSUE TITLE
Make conrod agnostic over graphics API 

### ISSUE THREAD

**Issue by masche842** [PistonDevelopers/conrod#658]
As discussed in #626 I am experimenting with backend implementations.

Here
https://github.com/masche842/conrod_backends
https://github.com/masche842/conrod

you can find a first sketch of the simplest possible backend implementation that does nothing more than printing out in clear text what should be rendered (like "draw rect at xy with color").

I copied over stuff from piston-graphics to conrod in order to get rid of the dependency. I think much/some of the code is superfluous and can be removed. But before I dig deeper into this I would like to hear your thoughts about it. Am I heading into the right direction? Or am I terribly wrong about my assumptions and there is a better way to do all this?

Maybe the most interesting thing is rendering text. How much implementation should be done in conrod itself? How abstract should the interface be? I.e. a backend where only monospaced fonts are used that are stored as fixed size byte arrays (see https://github.com/achimdoebler/UGUI/blob/master/ugui.c#L66)  probably doesn't need a GlyphCache at all. (On the other hand: Maybe in that case conrod is not the right weapon for the fight anyway. To be evaluated: how slim and fast can it be with a custom backend?)

**Comment by bvssvni** [comment]
Nice of you to think of this!

I think there are some problems that needs to be addressed with a simplified immediate mode interface:

- Restricts the user from what kind of widgets can be made
- Retained or exotic (for example built on top of texture arrays) won't have the same space for optimization

The second point was a part of motivation for considering alternative designs, but so far we don't know how this would look like.

Not sure about the direction with a simplified immediate mode interface will bring significant benefits. Of course, in theory there might be an optimal design out there, but it is the cost of moving there that I'm worried about.

**Comment by mitchmindtree** [comment]
@masche842 I think a good first step would be to highlight exactly what the `Graphics` trait lacks in expressivity and exactly what could be simplified.

I recently had a much closer look at the `Graphics` trait, and I realised that it is already much simpler than I'd assumed (and simpler than the "simplified" API I had in mind for conrod) - it only requires [4 methods](http://docs.piston.rs/graphics/graphics/trait.Graphics.html), which automatically give you the ability to be able to draw any convex shape or texture, and in turn text and images also.

Any of the shape or image methods can then be overridden manually if higher performance is necessary, and `CharacterCache` can be implemented for custom types if higher performance is necessary for rendering text.

Also, there are now better docs [on the src itself](https://github.com/PistonDevelopers/graphics/blob/master/src/graphics.rs) - it doesn't look like the online docs have updated yet.

At the moment, I'm leaning towards simply continuing to use the `Graphics` trait while ensuring that all necessary types and traits are exposed from the conrod backend module and all docs are inlined, perhaps with an example implementation in the `examples` directory.

What you mention about text rendering does sound interesting though. It looks like your example is a conrod implementation for a terminal style console backend? I hadn't originally considered this as a conrod use-case as it seems to diverge from GUI to TUI. I share @bvssvni's concern for the cost involved in getting this working, however I'm open to any ideas you have for making this possible :)

What are your thoughts on all this? Also, I'd love to hear more details about what you're hoping to use conrod for if you get the chance - maybe it'd give us a clearer idea of what you're going for in regards to the backend API?

**Comment by masche842** [comment]
@mitchmindtree While implementing (read: copy-and-pasting over) my supposed-to-be-simpler trait I also realized that the Graphics trait is already quite light-weight. I agree that continuing using it is the best way for now.

I didn't have a TUI in mind, I just wanted to point out what information is provided to a potential custom backend and needs to be handled there.

Regarding your last question: Currently I'm concerned that conrod may not be the right tool for the job. 
What I initially was looking for is a GUI Toolkit written in Rust that can handle small displays up to ~800x600 pixel (most of the time much smaller: 340x200 or even 128x64) and is performant enough to run on an embedded system (maybe even in the context of a unikernel).

I think the idea of an immediate mode GUI with a distinct global state fits in there. I also like conrod's way to reduce overhead introduced by this (by caching part of the widgets). But my concern is the additional stage of abstraction of converting primitives into triangles and back into primitives (as I don't have a GPU at hand) will be too expensive here. Text rendering in these context usually is handled by rendering monospaced characters directly from const arrays without the additional hassle of calculating widths and caching and so on.

Anyway, I will give it a try by implementing a custom backend for the graphics trait, as I have already scaffolded things (I got stuck converting between the coordinate systems). When initial benchmarks show promising results I will drop you a line. Otherwise I will either have to implement my own toolkit or go back to C and a well established library (last resort...).

Either way thanks @bvssvni and @mitchmindtree (and all other contributors) for your awesome work on conrod, I think it has huge potential!

As the initial "issue" is off the table I will close it now. Of course I'm still curious about your thoughts on my use case!
